### PR TITLE
Fix the handling of non-flat intermediate vectors in AverageAggregate

### PR DESCRIPTION
**Pull request by kagamiori**
Summary:
AverageAggregate::addIntermediateResults and addSingleGroupIntermediateResults
assume that the fields in the intermediate result RowVector are all flat. This is not true
when the companion functions, e.g., avg_merge and avg_merge_extract_double, are
used by themselves. This diff fixes this issue by letting AverageAggregate::addIntermediateResults
and addSingleGroupIntermediateResults decode the fields of the input vector.

The cost of decoding field vectors is negligible. Below is the benchmark result before and after this fix.
Before:
```
============================================================================
[...]gates/benchmarks/AverageBenchmark.cpp     relative  time/iter   iters/s
============================================================================
flat                                                      194.17ms      5.15
constant                                        152.85%   127.03ms      7.87
dictionary                                      78.342%   247.85ms      4.03
```

After:
```
============================================================================
[...]gates/benchmarks/AverageBenchmark.cpp     relative  time/iter   iters/s
============================================================================
flat                                                      192.22ms      5.20
constant                                        149.39%   128.67ms      7.77
dictionary                                      77.426%   248.26ms      4.03
```

Differential Revision: D47422668

### CODE REVIEW

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp
Comment (bikramSingh91) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:273]: Wonder if it makes sense to have (decodedPartial_.isConstantMapping()) as the top level check. This will work for both flat and encoded children as we dont need to put the sum and count vector in a decoded vector to read the values. only 1 value needs to be read, therefore traversing multiple layers once for each should be ok and faster because we dont traverse all rows when creating a decoded vector.

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp
Comment (bikramSingh91) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:302]: I wonder if its possible to create a utility function for this code since its being repeated quite often with minor differences. Those differences can be enclosed within constexpr checks and templated. What do you think?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp
Comment (bikramSingh91) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:239]: do we have a benchmark for this agg? I wonder if we can just structure the cases like 
case1 : (decodedPartial_.isConstantMapping())
case2: (decodedPartial_.isIdentityMapping())
case3 : ELSE

Then create decodedVector for the children in case 2 and 3.
The reason I ask about benchmark is because creating a decodedVector for flat should be fairly lightweight as i directly points to the values and nulls buffers without doing any computation. So we can skip the special case for when both children are flat and ensure using the benchmark that we do not regress in perf.

File: velox/functions/prestosql/aggregates/tests/AverageAggregationTest.cpp
Comment (bikramSingh91) [inline velox/functions/prestosql/aggregates/tests/AverageAggregationTest.cpp:537]: should we check for the case where either one of children is flat? can there be multiple encoding layers?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp
Comment (kagamiori) [reply velox/functions/prestosql/aggregates/AverageAggregate.cpp]: According to offline discussion, I updated the code to decode child vectors in all cases.

File: velox/functions/prestosql/aggregates/tests/AverageAggregationTest.cpp
Comment (bikramSingh91) [inline velox/functions/prestosql/aggregates/tests/AverageAggregationTest.cpp:565]: can you add a small comment about the composition of rowInDict and row2 something like Dictionary(ROW(Flat, Flat)) and ROW(Lazy, Dictionary(Constant(flat)). Maybe also mention in the test name that we are testing lazy too, not just non-flat

**Review by bikramSingh91 — APPROVED** [review]
Looks good. Approving PR to unblock, but please address the new comments that i posted before merging.

**Pull request closed without merging**
